internal/webhook/webhook.go

```diff
@@ -0,0 +1,154 @@
+	config "github.com/jetstack/cert-manager/internal/apis/config/webhook"
+	metainstall "github.com/jetstack/cert-manager/internal/apis/meta/install"
+	"github.com/jetstack/cert-manager/internal/plugin"
+	logf "github.com/jetstack/cert-manager/pkg/logs"
+	"github.com/jetstack/cert-manager/pkg/webhook/admission"
+	"github.com/jetstack/cert-manager/pkg/webhook/admission/initializer"
+	"github.com/jetstack/cert-manager/pkg/webhook/authority"
+	"github.com/jetstack/cert-manager/pkg/webhook/handlers"
+	"github.com/jetstack/cert-manager/pkg/webhook/server"
+	"github.com/jetstack/cert-manager/pkg/webhook/server/tls"
+)
+
+var conversionHook handlers.ConversionHook = handlers.NewSchemeBackedConverter(logf.Log, Scheme)
+
+type ServerOption func(*server.Server)
+
+// WithConversionHandler allows you to override the handler for the `/convert`
+// endpoint in tests.
+func WithConversionHandler(handler handlers.ConversionHook) ServerOption {
+	return func(s *server.Server) {
+		s.ConversionWebhook = handler
+	}
+}
+
+// NewCertManagerWebhookServer creates a new webhook server configured with all cert-manager
+// resource types, validation, defaulting and conversion functions.
+func NewCertManagerWebhookServer(log logr.Logger, _ options.WebhookFlags, opts config.WebhookConfiguration, optionFunctions ...ServerOption) (*server.Server, error) {
+	restcfg, err := clientcmd.BuildConfigFromFlags(opts.APIServerHost, opts.KubeConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	cl, err := kubernetes.NewForConfig(restcfg)
+	if err != nil {
+		return nil, fmt.Errorf("error creating kubernetes client: %s", err)
+	}
+
+	// Set up the admission chain
+	admissionHandler, err := buildAdmissionChain(log, cl)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &server.Server{
+		ListenAddr:        fmt.Sprintf(":%d", *opts.SecurePort),
+		HealthzAddr:       fmt.Sprintf(":%d", *opts.HealthzPort),
+		EnablePprof:       opts.EnablePprof,
+		PprofAddr:         opts.PprofAddress,
+		CertificateSource: buildCertificateSource(log, opts.TLSConfig, restcfg),
+		CipherSuites:      opts.TLSConfig.CipherSuites,
+		MinTLSVersion:     opts.TLSConfig.MinTLSVersion,
+		ValidationWebhook: admissionHandler,
+		MutationWebhook:   admissionHandler,
+		ConversionWebhook: conversionHook,
+		Log:               log,
+	}
+	for _, fn := range optionFunctions {
+		fn(s)
+	}
+	return s, nil
+}
+
+func buildAdmissionChain(log logr.Logger, client kubernetes.Interface) (*admission.RequestHandler, error) {
+	// Set up the admission chain
+	pluginHandler := admission.NewPlugins(log, Scheme)
+	plugin.RegisterAllPlugins(pluginHandler)
+	authorizer, err := authorizerfactory.DelegatingAuthorizerConfig{
+		SubjectAccessReviewClient: client.AuthorizationV1(),
+		// cache responses for 1 second
+		AllowCacheTTL: time.Second,
+		DenyCacheTTL:  time.Second,
+		WebhookRetryBackoff: &wait.Backoff{
+			Duration: time.Second,
+			Factor:   2,
+			Jitter:   0.2,
+			Steps:    2,
+			Cap:      time.Second * 5,
+		},
```

Comment:
This represents a change from how the approval plugin used to work - previously a live SubjectAccessReview using the k8s client would be performed. Now, we use a caching authorizer which will cache responses to avoid having multiple lookups (with a fixed expiry of 1s for now) as well as retrying requests if they fail.